### ISSUE TITLE
[Question] How to validate an array of enum values

### ISSUE THREAD

**Issue by cloakedch** [typestack/class-validator#159]
How would one validate a field that contains only certain string values?

For instance validate the following enum:
<pre><code>
export enum Permissions {
  READ_USER_SELF = "READ_USER_SELF",
  READ_USER_ALL = "READ_USER_ALL"
};
</code></pre>

My Dto contains:

<pre><code>
export class UserDto {
  @IsEmail()
  readonly email: string;

  @IsOptional()
  @IsArray()
  readonly permissions: Permissions[]; <- How to validate this?
}
</code></pre>

So in other words: How can I validate that an object contains an array of predefined strings?

**Comment by oschweitzer** [comment]
Does  `@IsIn(["READ_USER_SELF", "READ_USER_ALL"])` work for you ?

**Comment by NoNameProvided** [comment]
As @oschweitzer wrote, the best way of doing this with the `@IsEnum` decorator. You can see in the [tests][1] how you can use it.

[1]: https://github.com/typestack/class-validator/blob/eb6ef1cfde6f6eda61b67dfa6de155580bee0232/test/functional/validation-functions-and-decorators.spec.ts#L689-L767

**Comment by cloakedch** [comment]
Thanks for both answers. However I checked them already before.

The @IsEnum approach doesn't work, as the validation of an array of enum values fails (even when the right values are submitted).

The @IsIn doesn't work either for arrays. It works perfectly if, in my given example, I had only one enum value. However in validation, the whole array is validated against single enum values.

Anyway, I solved it using a custom validator using @Validate. In there, I also do a duplication test, so not two enum values are present. That even enhanced the solution.

**Comment by pablocarreraest** [comment]
@NoNameProvided how isEnum works if a need to validate a string array?

I see that is Enum have to pass a entity value..?

or like @IsEnum(['opt1', 'opt2', 'opt3']) works right?

**Comment by NoNameProvided** [comment]
Ohh, I misunderstood you. You can only use enums if you expect a single value from an array of values. If you expect an array and want to validate every item in that array against a list of possible values then you need to use a custom validator (`@Validate`) for that as you said.

**Comment by fwoelffel** [comment]
How about
```js
export class UserDto {
  @IsEmail()
  readonly email: string;

  @IsOptional()
  @IsEnum(Permissions, { each: true })
  readonly permissions: Permissions[]; <- How to validate this?
}
```

I didn't try to execute this, but according to the docs it should work. If it doesn't we should submit a PR to fix this.

**Comment by cloakedch** [comment]
Looks like this works indeed @fwoelffel 👍 
I stick to my solution though, as I prefer to have a duplication test as well in the array.

**Comment by fwoelffel** [comment]
@cloakedch There is also `@ArrayUnique()` if you don't want to maintain an extraneous custom validator :wink:

**Comment by cloakedch** [comment]
@fwoelffel Thank you, seems to work indeed! I must have missed that one :-O
